active/0024-message-validation.md

```diff
@@ -0,0 +1,132 @@
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but they do not interfere MQTT message pub/sub.
+For example, if there is client subscribed to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
+
+### Message filter
+
+Message filtering can be achieved by confugring message validation rule to drop invalid messages.
+
+## Design
+
+This data validation step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Make use of the rules engine SQL, but without the `FROM` clause as it's always the MQTT message being th input.
+
+Each data validation rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+### Message Validation
+
+```
+validation1 {
+    enable = true # or 'false'
+    type = validation
+    description = "drop message if it is not compabitble with my avro schema or if payload.value is less tha 0"
+    topics = "t/#" # or topics = ["t/1", "t/2"]
+    check_mode = any_pass # or all_pass
+    failure_action = disconnect # (disconnect also implies 'dorp') or 'drop' to only drop the message
+    checks = [
+        {
+            # message payload payload to be validated against JSON schema
+            type = json # can also be 'avro' or 'protobuf'
+            schema = "my-json-schema-registration-id"
+        }
+        {
+            # SQL evaluates to empty result indicates validation falure
+            # in this example, if payload.value is less than 0, the message is considiered invalid
+            type = sql
+            sql = "SELECT * WHERE payload.value < 0"
+        }
+    ]
+}
+```
+
+If there are more than one validation matched for one message, all validations should be executed.
+For example, if one is configured with `topics = "t/#"` and another with `topics = "t/1"`,
+when a message is published to `t/1`, the both validations should be triggered (without ordering guarantee).
+
+## Configuration Changes
+
+A new config root named `message_validation` is to be added.
+The value of this config is a `Map` which has the validation ID as key, and configs as value.
+
+## APIs
+
+- GET /message_validations
```

Comment:
Supports filtering by topic and schema：
```suggestion
- GET /message_validations?topic=foo&topic_match=t/#&schema=foo&type=json
```

Reply:
good idea, however not quite sure what's the purpose of `topic` in conjunction with `topic_match` by `&`.